iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnBuffers.cpp

```diff
@@ -535,84 +536,30 @@ LogicalResult PadOpConversion::apply(
-
-  SmallVector<Type, 2> resultTypes = {};
-  SmallVector<Value, 2> linalgOpArgs = {adaptor.operand()};
-  if (!paddingConstVal) linalgOpArgs.push_back(adaptor.padding_value());
-  linalgOpArgs.push_back(resultBuffers[0]);
-  auto linalgOp = rewriter.create<linalg::IndexedGenericOp>(
-      loc, resultTypes, linalgOpArgs,
-      rewriter.getI64IntegerAttr(linalgOpArgs.size() - 1),  // args_in
-      rewriter.getI64IntegerAttr(1),                        // args_out
-      rewriter.getArrayAttr(indexingMaps),
-      getParallelAndReductionIterAttrs(rewriter, rank, /*nReduction=*/0),
-      /*doc=*/nullptr, /*library_call=*/nullptr);
-
-  // Add a block to the region.
-  auto *region = &linalgOp.region();
-  auto *block = rewriter.createBlock(region, region->end());
-  SmallVector<Type, 4> bodyArgTypes;
-  bodyArgTypes.append(rank, rewriter.getIndexType());
-  bodyArgTypes.append(linalgOpArgs.size(), operandType.getElementType());
-  block->addArguments(bodyArgTypes);
-  rewriter.setInsertionPointToEnd(block);
-
-  // If the `index` of the result at a particular dimension i, is d_i, check if
-  //
-  // (d_i >= edge_padding_low[i]) &&
-  // (d_i < (edge_padding_low[i] + operand_shape[i])).
-  //
-  // If true, then use the value of the operand, otherwise use the padding
-  // value.
   const auto &edgePaddingLow = op.edge_padding_low();
-  const auto &edgePaddingHigh = op.edge_padding_high();
+  const auto &interiorPadding = op.interior_padding();
+  SmallVector<Value, 3> offsets, sizes, strides;
+  for (auto it : llvm::enumerate(llvm::zip(edgePaddingLow, interiorPadding))) {
+    Value startIndex = rewriter.create<ConstantIndexOp>(
+        loc, std::get<0>(it.value()).getZExtValue());
+    offsets.push_back(startIndex);
+    Value size = rewriter.create<DimOp>(loc, resultBuffers[0], it.index());
+    sizes.push_back(size);
+    Value stride = rewriter.create<ConstantIndexOp>(
+        loc, std::get<1>(it.value()).getZExtValue() + 1);
+    strides.push_back(stride);
+  }
 
-  Type indexType = rewriter.getIndexType();
-  Value cond = nullptr;
-  auto applyAndOp = [&](Value val) {
-    cond = cond ? rewriter.create<AndOp>(loc, cond, val) : val;
-  };
-  for (int i = 0; i < rank; ++i) {
-    Value dim = block->getArgument(i);
-    int64_t paddingLow = edgePaddingLow.getValue<IntegerAttr>(i).getInt();
-    int64_t paddingHigh = edgePaddingHigh.getValue<IntegerAttr>(i).getInt();
-    auto low = rewriter.create<ConstantOp>(
-        loc, indexType, rewriter.getIntegerAttr(indexType, paddingLow));
-
-    // d_i < (edge_padding_low[i] + operand_shape[i])
-    if (paddingLow != 0 && paddingHigh != 0) {
-      auto operandExtent = rewriter.create<DimOp>(loc, adaptor.operand(), i);
-      auto bound = rewriter.create<AddIOp>(loc, low, operandExtent);
-      auto checkUb =
-          rewriter.create<CmpIOp>(loc, CmpIPredicate::slt, dim, bound);
-      applyAndOp(checkUb);
-    }
+  // The pass for splitting dispatch function for vulkan requires no other ops
```

Comment:
That pass should be relaxed to allow "metadata" ops in the middle. Technically this pass should not be concerning its downstream users.. But it's fine for now. :)

Reply:
Cool, I added a TODO here. Once the condition is relaxed, we can move it to before where it's used.